<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU18_Infraestructuras_eventos</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 05. Agrupar datos de administración y empresas por distrito

* A partir de los datos puntuales, agregar por distrito censal y contar infraestructruras para hacer mapas de regiones. 
* Adicionalmente crear metadatos con descripción y agrupamiento de variables.








<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use

- {tcltk} para selección interactiva de archivos locales
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {dityr} para tranformar datos
- {janitor} para limpiar datos


In [1]:
library(readr)
library(dplyr)
library(tidyr)
library(janitor)



Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



Attaching package: ‘janitor’


The following objects are masked from ‘package:stats’:

    chisq.test, fisher.test




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [3]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [3]:
iFile <- "CU_18_05_08_adm_distritos.csv"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_18_05_08_adm_distritos.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [4]:
data <- read_csv(file_data)


Rows: 120280 Columns: 7
── Column specification ─────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (5): Grupo, Tipo, ETIQUETA, CMUN, CDIS
dbl (2): X, Y

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [5]:
glimpse(data)

Rows: 120,280
Columns: 7
$ Grupo    <chr> "Administración pública", "Administración pública", "Administ…
$ Tipo     <chr> "Agencia Tributaria", "Agencia Tributaria", "Agencia Tributar…
$ ETIQUETA <chr> "Administración de la Agencia Tributaria Oficina Central", "A…
$ X        <dbl> -3.698678, -3.712513, -3.745598, -3.690993, -3.672477, -3.710…
$ Y        <dbl> 40.45554, 40.44553, 40.37049, 40.41795, 40.45483, 40.48316, 4…
$ CMUN     <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079"…
$ CDIS     <chr> "06", "07", "11", "03", "05", "08", "04", "14", "20", "12", "…


Muestra de datos:

In [6]:
slice_head(data, n = 5)

Grupo                  Tipo              
1 Administración pública Agencia Tributaria
2 Administración pública Agencia Tributaria
3 Administración pública Agencia Tributaria
4 Administración pública Agencia Tributaria
5 Administración pública Agencia Tributaria
  ETIQUETA                                                                      
1 Administración de la Agencia Tributaria Oficina Central                       
2 Administración de la Agencia Tributaria Guzmán el Bueno Administración Central
3 Administración de la Agencia Tributaria  Suroeste                             
4 Administración de la Agencia Tributaria Montalbán                             
5 Administración de la Agencia Tributaria Ciudad Lineal                         
  X         Y        CMUN CDIS
1 -3.698678 40.45554 079  06  
2 -3.712513 40.44553 079  07  
3 -3.745598 40.37049 079  11  
4 -3.690993 40.41795 079  03  
5 -3.672477 40.45483 079  05

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Infraestructuras

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

#### Data Transform

* Contar número de infraestructuras de cada tipo por distrito
* Extender en columnas para caracterizar distritos

In [7]:
tdata_01 <- data |> 
  count(CMUN, CDIS, Tipo) |> 
  pivot_wider(names_from = Tipo,
              values_from = n, 
              values_fill = 0) 

In [8]:
glimpse(tdata_01)

Rows: 246
Columns: 49
$ CMUN                                                          <chr> "001", "…
$ CDIS                                                          <chr> "01", "0…
$ `Ayuntamientos, Consejerías, Ministerios, etc.`               <int> 1, 1, 1,…
$ `Sanidad y servicios sociales`                                <int> 1, 8, 0,…
$ `Actividades administrativas`                                 <int> 0, 13, 0…
$ `Actividades asociativas y hogares`                           <int> 0, 2, 0,…
$ `Actividades profesionales`                                   <int> 0, 11, 1…
$ `Administraciones públicas`                                   <int> 0, 2, 2,…
$ Alimentación                                                  <int> 0, 2, 0,…
$ `Centros educativos no universitarios - Centros públicos`     <int> 0, 3, 0,…
$ `Comercio mayorista`                                          <int> 0, 34, 0…
$ `Comercio minorista`                                          <int> 0, 22, 0…
$ Construcción    

In [9]:
tdata_01 |> slice_head(n = 5)

CMUN CDIS Ayuntamientos, Consejerías, Ministerios, etc.
1 001  01   1                                            
2 002  01   1                                            
3 003  01   1                                            
4 004  01   1                                            
5 005  01   1                                            
  Sanidad y servicios sociales Actividades administrativas
1  1                            0                         
2  8                           13                         
3  0                            0                         
4 10                            2                         
5 42                           36                         
  Actividades asociativas y hogares Actividades profesionales
1  0                                 0                       
2  2                                11                       
3  0                                 1                       
4  2                                 3                       
5 14                                56                       
  Administraciones públicas Alimentación
1 0                         0           
2 2                         2           
3 2                         0           
4 2                         0           
5 5                         4           
  Centros educativos no universitarios - Centros públicos ⋯
1 0                                                       ⋯
2 3                                                       ⋯
3 0                                                       ⋯
4 4                                                       ⋯
5 8                                                       ⋯
  Galerías de alimentación Grandes superficies especializadas Hipermercados
1 0                        0                                  0            
2 0                        0                                  0            
3 0                        0                                  0            
4 0                        0                                  0            
5 3                        1                                  1            
  Mercados de abastos Agricultura y ganadería
1 0                   0                      
2 0                   0                      
3 0                   0                      
4 0                   0                      
5 1                   0                      
  Centros educativos no universitarios - Servicios educativos Seguridad Social
1 0                                                           0               
2 0                                                           0               
3 0                                                           0               
4 0                                                           0               
5 0                                                           0               
  Centros comerciales Oficinas de empleo Embajadas y consulados
1 0                   0                  0                     
2 0                   0                  0                     
3 0                   0                  0                     
4 0                   0                  0                     
5 0                   0                  0

#### Data Extract




* Extraer nombres de columnas


In [44]:
tdata_02 <- data.frame(desc_var = colnames(tdata_01)[3:ncol(tdata_01)])

In [45]:
tdata_02

desc_var                                                 
1  ayuntamientos_consejerias_ministerios_etc                
2  sanidad_y_servicios_sociales                             
3  actividades_administrativas                              
4  actividades_asociativas_y_hogares                        
5  actividades_profesionales                                
6  administraciones_publicas                                
7  alimentacion                                             
8  centros_educativos_no_universitarios_centros_publicos    
9  comercio_mayorista                                       
10 comercio_minorista                                       
11 construccion                                             
12 educacion                                                
13 hosteleria                                               
14 industria_quimica_y_refino                               
15 informacion_y_comunicaciones                             
16 maquinaria_industrial                                    
17 material_de_transporte                                   
18 material_electrico_y_electronico                         
19 mercadillos                                              
20 metalicas_basicas_e_intermedias                          
21 mineria_y_suministros                                    
22 otras_manufactureras                                     
23 otros_servicios_bancos                                   
24 papel_y_graficas                                         
25 servicios_financieros                                    
26 servicios_personales                                     
27 servicios_recreativos                                    
28 textil_confeccion_y_calzado                              
29 transporte_y_almacenamiento                              
30 actividades_inmobiliarias                                
31 centros_educativos_no_universitarios_centros_privados    
32 industria_no_metalica                                    
33 administracion_de_justicia                               
34 agencia_tributaria                                       
35 campus_universitarios                                    
36 centros_educativos_universitarios                        
37 colegios_mayores                                         
38 galerias_de_alimentacion                                 
39 grandes_superficies_especializadas                       
40 hipermercados                                            
41 mercados_de_abastos                                      
42 agricultura_y_ganaderia                                  
43 centros_educativos_no_universitarios_servicios_educativos
44 seguridad_social                                         
45 centros_comerciales                                      
46 oficinas_de_empleo                                       
47 embajadas_y_consulados

Data Transform

* Limpiar nombres de columnas

In [12]:
tdata_01 <- tdata_01 |> 
  clean_names()

In [13]:
glimpse(tdata_01)

Rows: 246
Columns: 49
$ cmun                                                      <chr> "001", "002"…
$ cdis                                                      <chr> "01", "01", …
$ ayuntamientos_consejerias_ministerios_etc                 <int> 1, 1, 1, 1, …
$ sanidad_y_servicios_sociales                              <int> 1, 8, 0, 10,…
$ actividades_administrativas                               <int> 0, 13, 0, 2,…
$ actividades_asociativas_y_hogares                         <int> 0, 2, 0, 2, …
$ actividades_profesionales                                 <int> 0, 11, 1, 3,…
$ administraciones_publicas                                 <int> 0, 2, 2, 2, …
$ alimentacion                                              <int> 0, 2, 0, 0, …
$ centros_educativos_no_universitarios_centros_publicos     <int> 0, 3, 0, 4, …
$ comercio_mayorista                                        <int> 0, 34, 0, 3,…
$ comercio_minorista                                        <int> 0, 22, 0, 10…
$ construccion    

* Completar metadatos

In [46]:
tdata_02 <- tdata_02 |>
    mutate(nombre_var = colnames(tdata_01)[3:ncol(tdata_01)]) |>
    left_join(data |> count(Grupo, Tipo), 
                by = c("desc_var" = "Tipo")) 


In [47]:
glimpse(tdata_02)

Rows: 47
Columns: 4
$ desc_var   <chr> "ayuntamientos_consejerias_ministerios_etc", "sanidad_y_ser…
$ nombre_var <chr> "ayuntamientos_consejerias_ministerios_etc", "sanidad_y_ser…
$ Grupo      <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…
$ n          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,…


In [48]:
tdata_02

desc_var                                                 
1  ayuntamientos_consejerias_ministerios_etc                
2  sanidad_y_servicios_sociales                             
3  actividades_administrativas                              
4  actividades_asociativas_y_hogares                        
5  actividades_profesionales                                
6  administraciones_publicas                                
7  alimentacion                                             
8  centros_educativos_no_universitarios_centros_publicos    
9  comercio_mayorista                                       
10 comercio_minorista                                       
11 construccion                                             
12 educacion                                                
13 hosteleria                                               
14 industria_quimica_y_refino                               
15 informacion_y_comunicaciones                             
16 maquinaria_industrial                                    
17 material_de_transporte                                   
18 material_electrico_y_electronico                         
19 mercadillos                                              
20 metalicas_basicas_e_intermedias                          
21 mineria_y_suministros                                    
22 otras_manufactureras                                     
23 otros_servicios_bancos                                   
24 papel_y_graficas                                         
25 servicios_financieros                                    
26 servicios_personales                                     
27 servicios_recreativos                                    
28 textil_confeccion_y_calzado                              
29 transporte_y_almacenamiento                              
30 actividades_inmobiliarias                                
31 centros_educativos_no_universitarios_centros_privados    
32 industria_no_metalica                                    
33 administracion_de_justicia                               
34 agencia_tributaria                                       
35 campus_universitarios                                    
36 centros_educativos_universitarios                        
37 colegios_mayores                                         
38 galerias_de_alimentacion                                 
39 grandes_superficies_especializadas                       
40 hipermercados                                            
41 mercados_de_abastos                                      
42 agricultura_y_ganaderia                                  
43 centros_educativos_no_universitarios_servicios_educativos
44 seguridad_social                                         
45 centros_comerciales                                      
46 oficinas_de_empleo                                       
47 embajadas_y_consulados                                   
   nombre_var                                                Grupo n 
1  ayuntamientos_consejerias_ministerios_etc                 NA    NA
2  sanidad_y_servicios_sociales                              NA    NA
3  actividades_administrativas                               NA    NA
4  actividades_asociativas_y_hogares                         NA    NA
5  actividades_profesionales                                 NA    NA
6  administraciones_publicas                                 NA    NA
7  alimentacion                                              NA    NA
8  centros_educativos_no_universitarios_centros_publicos     NA    NA
9  comercio_mayorista                                        NA    NA
10 comercio_minorista                                        NA    NA
11 construccion                                              NA    NA
12 educacion                                                 NA    NA
13 hosteleria                                                NA    NA
14 industria_quimica_y_refino                                NA    NA
15 informacion_y_comunica

Si no aplica: Estos datos no requieren tareas de este tipo.

## Synthetic Data Generation

Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Estos datos no requieren tareas de este tipo.

## Open Data

Estos datos no requieren tareas de este tipo.

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

1. Infraestructuras agregadas por distrito

In [26]:
data_to_save <- tdata_01


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [30]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_09"
archivo <- "_01"
proper <- "_distritos_adm"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save, path_out_01)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [31]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_18_05_09_01_distritos_adm.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [32]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Metadatos de infraestructuras

In [49]:
data_to_save_02 <- tdata_02

In [50]:
archivo <- "_02"
proper <- "_admin_meta"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [51]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_18_05_09_02_admin_meta.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [38]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {readr}, {dplyr}, {tidyr} and {janitor} packages must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* tidyr 1.3.0
* dplyr 1.0.10
* janitor 2.1.0
* readr 2.1.3

### Data structures

#### Objeto `tdata_01`

- Tenemos 246 filas, una por distrito, con los recuentos de cada uno de los tipos de infraestructuras (12)



In [41]:
glimpse(tdata_01)

Rows: 246
Columns: 49
$ cmun                                                      <chr> "001", "002"…
$ cdis                                                      <chr> "01", "01", …
$ ayuntamientos_consejerias_ministerios_etc                 <int> 1, 1, 1, 1, …
$ sanidad_y_servicios_sociales                              <int> 1, 8, 0, 10,…
$ actividades_administrativas                               <int> 0, 13, 0, 2,…
$ actividades_asociativas_y_hogares                         <int> 0, 2, 0, 2, …
$ actividades_profesionales                                 <int> 0, 11, 1, 3,…
$ administraciones_publicas                                 <int> 0, 2, 2, 2, …
$ alimentacion                                              <int> 0, 2, 0, 0, …
$ centros_educativos_no_universitarios_centros_publicos     <int> 0, 3, 0, 4, …
$ comercio_mayorista                                        <int> 0, 34, 0, 3,…
$ comercio_minorista                                        <int> 0, 22, 0, 10…
$ construccion    

## Objeto `tdata_02`

- Tenemos 12 filas, una por tipo de infraestructura, con sus metadatos



In [40]:
glimpse(tdata_02)

Rows: 47
Columns: 6
$ desc_var   <chr> "Ayuntamientos, Consejerías, Ministerios, etc.", "Sanidad y…
$ nombre_var <chr> "ayuntamientos_consejerias_ministerios_etc", "sanidad_y_ser…
$ Grupo.x    <chr> "Administración pública", "Colectivo empresarial por tamaño…
$ n.x        <int> 245, 6422, 7265, 3103, 10696, 965, 806, 1983, 7391, 16908, …
$ Grupo.y    <chr> "Administración pública", "Colectivo empresarial por tamaño…
$ n.y        <int> 245, 6422, 7265, 3103, 10696, 965, 806, 1983, 7391, 16908, …


#### Observaciones generales sobre los datos

* No aplica

### Consideraciones para despliegue en piloto

* Utilizar los metadatos para etiquetar gráficos y otras salidas.

### Consideraciones para despliegue en producción

* No aplica

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han agrupado las infraesctructuras por distrito
- Se han calculado el número de infraestructuras por tipo en cada distrito

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Los datos se unirán a los de infraestructuras para ajustar los modelos

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código